File: packages/template-go/value/value.go
// Package value defines the value types, safe strings, truthiness, equality and ordering
// of docs/spec/data-model.md and docs/spec/expressions.md.
package value

import (
	"regexp"
	"strconv"
	"strings"
	"unicode/utf8"
)

// Value is one of nil, bool, float64, string, SafeString, List or *OrderedMap.
type Value = any

// List is an ordered sequence of values.
type List = []Value

// SafeString is a string that the echo tag writes without escaping (VAL-6).
type SafeString struct {
	Text string
}

// Type is a value type name (FUN-19).
type Type string

// Value types.
const (
	TypeNull   Type = "null"
	TypeBool   Type = "bool"
	TypeNumber Type = "number"
	TypeString Type = "string"
	TypeList   Type = "list"
	TypeMap    Type = "map"
)

// TypeOf returns the type name of a value.
func TypeOf(v Value) Type {
	switch v.(type) {
	case nil:
		return TypeNull
	case bool:
		return TypeBool
	case float64:
		return TypeNumber
	case string, SafeString:
		return TypeString
	case List:
		return TypeList
	case *OrderedMap:
		return TypeMap
	}
	return TypeNull
}

// IsString reports whether a value is a string or a safe string.
func IsString(v Value) bool {
	switch v.(type) {
	case string, SafeString:
		return true
	}
	return false
}

// TextOf returns the text of a string or safe string; ok is false for other values.
func TextOf(v Value) (text string, ok bool) {
	switch s := v.(type) {
	case string:
		return s, true
	case SafeString:
		return s.Text, true
	}
	return "", false
}

// CodePointLength returns the number of Unicode code points (VAL-5).
func CodePointLength(s string) int {
	return utf8.RuneCountInString(s)
}

// CompareCodePoints compares two strings by code point sequence (VAL-5).
func CompareCodePoints(a, b string) int {
	return strings.Compare(a, b)
}

// IsTruthy implements EXP-33.
func IsTruthy(v Value) bool {
	switch x := v.(type) {
	case nil:
		return false
	case bool:
		return x
	case float64:
		return x != 0
	case string:
		return x != ""
	case SafeString:
		return x.Text != ""
	case List:
		return len(x) > 0
	case *OrderedMap:
		return x.Len() > 0
	}
	return false
}

var numberGrammar = regexp.MustCompile(`^[+-]?([0-9]+(\.[0-9]*)?|\.[0-9]+)([eE][+-]?[0-9]+)?$`)

func trimASCII(s string) string {
	return strings.Trim(s, " \t\r\n")
}

// ParseNumericString returns the numeric value of a string under EXP-23, or false.
func ParseNumericString(s string) (float64, bool) {
	t := trimASCII(s)
	if !numberGrammar.MatchString(t) {
		return 0, false
	}
	f, err := strconv.ParseFloat(t, 64)
	if err != nil {
		if numErr, ok := err.(*strconv.NumError); ok && numErr.Err == strconv.ErrRange {
			return 0, false
		}
		return 0, false
	}
	return f, true
}

// LooseEquals implements EXP-34 and EXP-35.
func LooseEquals(a, b Value) bool {
	ta, tb := TypeOf(a), TypeOf(b)
	if ta == tb {
		return sameTypeEquals(a, b, ta)
	}
	if ta == TypeNumber && tb == TypeString {
		text, _ := TextOf(b)
		n, ok := ParseNumericString(text)
		return ok && n == a.(float64)
	}
	if ta == TypeString && tb == TypeNumber {
		text, _ := TextOf(a)
		n, ok := ParseNumericString(text)
		return ok && n == b.(float64)
	}
	return false
}

// StrictEquals implements EXP-37.
func StrictEquals(a, b Value) bool {
	ta := TypeOf(a)
	return ta == TypeOf(b) && sameTypeEquals(a, b, ta)
}

func sameTypeEquals(a, b Value, t Type) bool {
	switch t {
	case TypeNull:
		return true
	case TypeBool:
		return a.(bool) == b.(bool)
	case TypeNumber:
		return a.(float64) == b.(float64)
	case TypeString:
		x, _ := TextOf(a)
		y, _ := TextOf(b)
		return x == y
	case TypeList:
		la, lb := a.(List), b.(List)
		if len(la) != len(lb) {
			return false
		}
		for i := range la {
			if !LooseEquals(la[i], lb[i]) {
				return false
			}
		}
		return true
	case TypeMap:
		ma, mb := a.(*OrderedMap), b.(*OrderedMap)
		if ma.Len() != mb.Len() {
			return false
		}
		for _, key := range ma.Keys() {
			other, ok := mb.Get(key)
			if !ok || !LooseEquals(ma.MustGet(key), other) {
				return false
			}
		}
		return true
	}
	return false
}

// Compare implements EXP-38; ok is false when the pair has no order.
func Compare(a, b Value) (order int, ok bool) {
	if x, isNum := a.(float64); isNum {
		if y, isNum2 := b.(float64); isNum2 {
			switch {
			case x < y:
				return -1, true
			case x > y:
				return 1, true
			}
			return 0, true
		}
		return 0, false
	}
	if x, isStr := TextOf(a); isStr {
		if y, isStr2 := TextOf(b); isStr2 {
			return CompareCodePoints(x, y), true
		}
	}
	return 0, false
}

File: packages/template-ts/src/value/value.ts
// Value types, safe strings, truthiness, equality and ordering as defined in
// docs/spec/data-model.md and docs/spec/expressions.md.

// A string that the echo tag writes without HTML escaping (VAL-6). The functions `raw` and
// `escape` produce one.
export class SafeString {
  // Wraps text that the echo tag writes as it is.
  constructor(readonly text: string) {}
}

/** A native assigned object whose public members and methods are visible to templates. */
export class NativeObject {
  constructor(readonly target: object) {}
}

// An ordered sequence of values (VAL-1).
export type ListValue = Value[];
// An ordered sequence of entries with string keys. Iteration follows insertion order (VAL-4).
export type MapValue = Map<string, Value>;
// A value that a template operates on (VAL-1).
export type Value = null | boolean | number | string | SafeString | ListValue | MapValue | NativeObject;

export type ValueType = 'null' | 'bool' | 'number' | 'string' | 'list' | 'map' | 'object';

export function typeOf(value: Value): ValueType {
  if (value === null) return 'null';
  if (typeof value === 'boolean') return 'bool';
  if (typeof value === 'number') return 'number';
  if (typeof value === 'string' || value instanceof SafeString) return 'string';
  if (Array.isArray(value)) return 'list';
  if (value instanceof NativeObject) return 'object';
  return 'map';
}

export function isString(value: Value): value is string | SafeString {
  return typeof value === 'string' || value instanceof SafeString;
}

export function textOf(value: string | SafeString): string {
  return typeof value === 'string' ? value : value.text;
}

// Number of Unicode code points of a string.
export function codePointLength(text: string): number {
  let count = 0;
  const iterator = text[Symbol.iterator]();
  while (!iterator.next().done) count++;
  return count;
}

export function codePoints(text: string): string[] {
  return Array.from(text);
}

// Compares two strings by code point sequence; returns a negative, zero or positive number.
export function compareCodePoints(a: string, b: string): number {
  const ia = a[Symbol.iterator]();
  const ib = b[Symbol.iterator]();
  for (;;) {
    const ra = ia.next();
    const rb = ib.next();
    if (ra.done && rb.done) return 0;
    if (ra.done) return -1;
    if (rb.done) return 1;
    const ca = (ra.value as string).codePointAt(0) as number;
    const cb = (rb.value as string).codePointAt(0) as number;
    if (ca !== cb) return ca < cb ? -1 : 1;
  }
}

// EXP-33: falsy values.
export function isTruthy(value: Value): boolean {
  if (value === null || value === false) return false;
  if (typeof value === 'number') return value !== 0;
  if (typeof value === 'string') return value.length > 0;
  if (value instanceof SafeString) return value.text.length > 0;
  if (Array.isArray(value)) return value.length > 0;
  if (typeof value === 'boolean') return value;
  if (value instanceof NativeObject) return true;
  return value instanceof Map ? value.size > 0 : true;
}

const numberGrammar = /^[+-]?([0-9]+(\.[0-9]*)?|\.[0-9]+)([eE][+-]?[0-9]+)?$/;

function trimAscii(text: string): string {
  let start = 0;
  let end = text.length;
  while (start < end && isAsciiSpace(text.charCodeAt(start))) start++;
  while (end > start && isAsciiSpace(text.charCodeAt(end - 1))) end--;
  return text.slice(start, end);
}

function isAsciiSpace(code: number): boolean {
  return code === 0x20 || code === 0x09 || code === 0x0d || code === 0x0a;
}

// Returns the numeric value of a string under the conversion grammar of EXP-23, or null.
export function parseNumericString(text: string): number | null {
  const trimmed = trimAscii(text);
  if (!numberGrammar.test(trimmed)) return null;
  const parsed = Number(trimmed);
  return Number.isFinite(parsed) ? parsed : null;
}

// EXP-34 and EXP-35.
export function looseEquals(a: Value, b: Value): boolean {
  const ta = typeOf(a);
  const tb = typeOf(b);
  if (ta === tb) return sameTypeEquals(a, b, ta);
  if (ta === 'number' && tb === 'string') {
    const n = parseNumericString(textOf(b as string | SafeString));
    return n !== null && n === (a as number);
  }
  if (ta === 'string' && tb === 'number') {
    const n = parseNumericString(textOf(a as string | SafeString));
    return n !== null && n === (b as number);
  }
  return false;
}

// EXP-37.
export function strictEquals(a: Value, b: Value): boolean {
  const ta = typeOf(a);
  return ta === typeOf(b) && sameTypeEquals(a, b, ta);
}

function sameTypeEquals(a: Value, b: Value, type: ValueType): boolean {
  switch (type) {
    case 'null':
      return true;
    case 'bool':
    case 'number':
      return a === b;
    case 'string':
      return textOf(a as string | SafeString) === textOf(b as string | SafeString);
    case 'list': {
      const la = a as ListValue;
      const lb = b as ListValue;
      return la.length === lb.length && la.every((item, index) => looseEquals(item, lb[index] as Value));
    }
    case 'map': {
      const ma = a as MapValue;
      const mb = b as MapValue;
      if (ma.size !== mb.size) return false;
      for (const [key, value] of ma) {
        if (!mb.has(key) || !looseEquals(value, mb.get(key) as Value)) return false;
      }
      return true;
    }
    case 'object':
      return a === b;
  }
}

// EXP-38: returns a negative, zero or positive number, or null when the pair has no order.
export function compareValues(a: Value, b: Value): number | null {
  if (typeof a === 'number' && typeof b === 'number') return a < b ? -1 : a > b ? 1 : 0;
  if (isString(a) && isString(b)) return compareCodePoints(textOf(a), textOf(b));
  return null;
}
